torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py

```diff
@@ -0,0 +1,130 @@
+"""JIT C++ strings into executables.
+
+1) Enable timeit and Callgrind collection for C++ snippets.
+2) Allow Timer.collect_callgrind to be used on earlier versions of PyTorch
+"""
+import atexit
+import os
+import re
+import shutil
+import textwrap
+import threading
+from types import ModuleType
+from typing import cast, List, Optional, Union, TYPE_CHECKING
+import uuid
+
+if TYPE_CHECKING:
+    from typing import Protocol
+else:
+    Protocol = object
+
+import torch
+from torch.utils import cpp_extension
+
+
+SOURCE_ROOT = os.path.split(os.path.abspath(__file__))[0]
+with open(os.path.join(SOURCE_ROOT, "timer_timeit_template.cpp"), "rt") as f:
+    CXX_TIMEIT_TEMPLATE = f.read()
+
+
+with open(os.path.join(SOURCE_ROOT, "timer_callgrind_template.cpp"), "rt") as f:
+    CXX_CALLGRIND_TEMPLATE = f.read()
+
+
+BUILD_ROOT = os.path.join(
+    torch._appdirs.user_cache_dir(appname="benchmark_utils_jit"),
+    f"build_{uuid.uuid4()}".replace("-", "")
+)
+os.makedirs(BUILD_ROOT, exist_ok=True)
+WRITE_SOURCE_LOCK = threading.Lock()
+atexit.register(shutil.rmtree, BUILD_ROOT)
```

Comment:
If these really are temporary files, it seems much better to actually put them in the system temporary directory, so they get cleared out even if your process segfaults or something.